File: quicksort.py
# list = [1,0,2,3,4,8,5,10,9,7,2,3,12,6,9]

low = 0
high = len(list) - 1

def quicksort(list, low, high):
    if (high - low) > 0:
        p = partition(list, low, high)
        quicksort(list, low, p-1)
        quicksort(list, p+1, high)

def partition(list, low, high):
    divider = low
    pivot = high

    for i in range(low, high):
        if list[i] < list[pivot]:
            list[i], list[divider] = list[divider], list[i]
            divider += 1
    list[divider], list[pivot] = list[pivot], list[divider]
    return divider
quicksort(list, low, high)

print(list)

File: QuickSort.cpp
/*
Quicksort
*/

#include <stdio.h>
#include <string.h>
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
#include <math.h>
#include <cstdlib>
#include <numeric>
#include <typeinfo>

#define SWAP(x,y,t) ((t)=(x), (x)=(y), (y)=(t))

using namespace std;

int list[10000000];
int partition(int list[], int left, int right)
{
    int pivot=list[left], tmp, low=left, high=right+1;

    do{
        do
        low++;
        while(low<=right && list[low]<pivot);

        do
        high--;
        while(high>=left && list[high]>pivot);
        if(low<high) SWAP(list[low], list[high], tmp);
    }while(low<high);

    SWAP(list[left], list[high], tmp);
    return high;
}
void quick_sort(int list[], int left, int right)
{
    if(left<right)
    {
        int q=partition(list, left, right);
        quick_sort(list, left, q-1);
        quick_sort(list, q+1, right);
    }
}

int main()
{
    int T;
    scanf("%d", &T);
    for (int i = 0; i < T; i++) {
        scanf("%d", &list[i]);
    }
    quick_sort(list, 0, T-1);
    for (int i = 0; i < T; i++)
        printf("%d\n", list[i]);
}
